### PR TITLE
streamingccl: add new logical package

### PR DESCRIPTION
**Pull request by stevendanna**
This adds a new LogicalReplication job and LogicalReplicationWriter
processor. This is a stub implementation that will be changed
substantially in the coming weeks.

The idea of the LogicalReplication job is that it can be used to
ingest data from one cluster into another while both clusters are
still accepting writes.

When two LogicalReplication jobs are started operating in opposite
directions on a pair of servers, the end effect will be an
active-active asynchronous replication process.

The overall architecture is similar to physical cluster replication.

Namely, for one flow, the destination cluster asks the source cluster
to establish a protected timestamp over the requested rows and to
return an ideal ingestion plan to the source.

The source then starts a set of workers each of which is responsible
for streaming a subset of the source cluster's keyspace. Rows are
returned over a pgwire connection.

The implementation here is contains almost none of the pieces required
to make this production ready.

It is missing, among many other things:

  1. Robust handling of schema descriptors to ensure that the user
  cannot make incompatible schema changes during replication.

  2. Correct handling of table's related by FK constraints.

  3. Easy-to-use SQL syntax.

  4. Correct last-write-wins semantics for DELETEs.

  5. Last-write-wins semantics without changes to the user's schema.

The goal of this PR is simply to get the job and processor in place so
that we can start filling out the production implementation.

Epic: none
Release note: None

### CODE REVIEW

**Review by msbutler — APPROVED** [review]
Looks production ready to me. Go ahead and merge now and address comments later if you'd like!

File: pkg/jobs/jobspb/jobs.proto [hunk 1]
```diff
@@ -207,6 +207,52 @@ message HistoryRetentionProgress {
...
+
+    repeated roachpb.Span source_spans = 2 [(gogoproto.nullable) = false];
+
+    // ReplicationStartTime is the initial timestamp from which the replication
+    // producer job will begin streaming MVCC revisions. This timestamp is picked
+    // once when the replication producer job is created, and is never updated
+    // through the lifetime of a replication stream. This will be the timestamp as
+    // of which each partition will perform its initial rangefeed scan on the
+    // source cluster.
+    util.hlc.Timestamp replication_start_time = 3 [(gogoproto.nullable) = false];
```
[the remark this answers is not in the document]
Comment (stevendanna) [reply pkg/jobs/jobspb/jobs.proto:234]: The second is the rule of thumb I was going by. When I move initial contact to planning, I'll move where these live.

**Comment by stevendanna** [comment]
bors r=msbutler

**Comment by nicktrav** [comment]
blathers backport 24.1

**Comment by nicktrav** [comment]
Ignore my derping 😂 ... wrong window.

**Pull request merged**